### PR TITLE
Add eth outbound queue and system pallets

### PR DESCRIPTION
**Pull request by tmpolaczyk**
Adds these pallets to Dancelight runtime:

```
        EthereumOutboundQueue: snowbridge_pallet_outbound_queue = 101,
        EthereumSystem: snowbridge_pallet_system = 103,
```

Requires some changes in polkadot-sdk, already merged in our stable2407 branch but not in polkadot upstream

### CODE REVIEW

File: solo-chains/runtime/dancelight/Cargo.toml
Comment (tmpolaczyk) [inline solo-chains/runtime/dancelight/Cargo.toml:146]: Remove from here, unused, used in tp-bridge

File: pallets/external-validator-slashes/src/mock.rs
Comment (girazoki) [inline pallets/external-validator-slashes/src/mock.rs:18]: FMT

File: pallets/external-validator-slashes/src/mock.rs
Comment (girazoki) [inline pallets/external-validator-slashes/src/mock.rs:246]: impled twice, one for MockErrOutboundQueue and the other for MockOkOutboundQueue. MockErrOutboundQueue is not used anymore right?

File: primitives/bridge/src/lib.rs
Comment (girazoki) [inline primitives/bridge/src/lib.rs:49]: add todo please

File: primitives/bridge/src/lib.rs
Comment (girazoki) [inline primitives/bridge/src/lib.rs:206]: FMT again

File: pallets/external-validator-slashes/src/lib.rs
Comment (girazoki) [inline pallets/external-validator-slashes/src/lib.rs:130]: I am wondering whether it makes sense to add our own `SendMessage` trait and implement it for things that implement their `SendMessage` (essentially being a wrapper here). This suggestion comes from the fact that we are implement their `sendMessage` in our mocks which does not make sense since we have our own custom validation

File: pallets/external-validator-slashes/src/lib.rs
Comment (girazoki) [inline pallets/external-validator-slashes/src/lib.rs:278]: we need to bench this

File: solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs
Comment (girazoki) [inline solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs:56]: put proper weight info benchmarked

File: solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs
Comment (girazoki) [inline solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs:50]: it seems to be this will be not enough, maybe we should double it.

in the case of 100 validators, 32 byte per validator account * 100 is already 3200...

File: solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs
Comment (girazoki) [inline solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs:144]: same thing with benchmarks

File: primitives/bridge/src/lib.rs
Comment (girazoki) [inline primitives/bridge/src/lib.rs:326]: are we doing the refund? based of this would it be easier to simply put a different origin than `snowbridge`?

File: solo-chains/runtime/dancelight/src/lib.rs
Comment (girazoki) [inline solo-chains/runtime/dancelight/src/lib.rs:998]: we could have snowbridge follow the previous path (through ethereumoutboundqueue) and come up with a new origin that routes everything through our `customProcessSnowbridgeMessage`... this origin is set in the sendMessage implementation of `EthereumOutboundQueue`. The solution to that would be to implement a differnt `SendMessage` implementation that sets it to the proper origin, and use this on external-validator-slashes

**Comment by girazoki** [comment]
Upstream changes look good, waiting for a potential PR opening and approval

File: primitives/bridge/src/lib.rs
Comment (Agusrodri) [inline primitives/bridge/src/lib.rs:237]:
```suggestion
```

File: solo-chains/runtime/dancelight/src/lib.rs
Comment (tmpolaczyk) [reply solo-chains/runtime/dancelight/src/lib.rs:1018]: Done, there are 2 different origins: Snowbridge for messages that are compatible with Snowbridge, and SnowbridgeTanssi for the Tanssi messages exclusively. Better name ideas welcome.

File: pallets/external-validator-slashes/src/mock.rs
Comment (tmpolaczyk) [reply pallets/external-validator-slashes/src/mock.rs:246]: No, I copied this from snowbridge, will remove it.

File: pallets/external-validator-slashes/src/lib.rs
Comment (tmpolaczyk) [inline pallets/external-validator-slashes/src/lib.rs:302]: Create error enums: EthereumValidateFail and EthereumDeliverFail

File: pallets/external-validator-slashes/src/lib.rs
Comment (tmpolaczyk) [reply pallets/external-validator-slashes/src/lib.rs:278]: It's only for testing, and ensure_root, what's the point of benchmarking it?

File: solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs
Comment (tmpolaczyk) [reply solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs:144]: I will do benchmarks when the PR is ready, as they are time consuming

File: solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs
Comment (tmpolaczyk) [reply solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs:50]: We decided not to do this and instead send the message in batches, correct?

File: pallets/external-validator-slashes/src/lib.rs
Comment (girazoki) [reply pallets/external-validator-slashes/src/lib.rs:282]: I dont think in the future it will be used only for testing, but rather, to inject a forced message if we for somehow skip it

File: solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs
Comment (girazoki) [reply solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs:50]: not in batches, but to send only the root. so yeah keep it like this

Comment (girazoki) [reply solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs:50]: ignore

**Review by girazoki — COMMENTED** [review]
My approval is implicit, but I refuse to tick the green button until we have benchmarked the instructions, just in case we miss them. But the pr looks good

**Review by girazoki — APPROVED** [review]
good job @tmpolaczyk

File: primitives/bridge/src/lib.rs
Comment (tmpolaczyk) [reply primitives/bridge/src/lib.rs:210]: Mainly because it adds some new imports and I wasn't sure if this is needed, so it would have been easier to remove it if it's all inside a module. I can move it to a separate file.

File: solo-chains/runtime/dancelight/src/lib.rs
Comment (tmpolaczyk) [reply solo-chains/runtime/dancelight/src/lib.rs:1746]: Ok I will clean up this part

File: primitives/bridge/src/lib.rs
Comment (ParthDesai) [reply primitives/bridge/src/lib.rs]: fixed

File: solo-chains/runtime/dancelight/src/lib.rs
Comment (ParthDesai) [reply solo-chains/runtime/dancelight/src/lib.rs:1744]: fixed

File: primitives/bridge/src/lib.rs
Comment (ParthDesai) [reply primitives/bridge/src/lib.rs]: fixed

File: solo-chains/runtime/dancelight/src/lib.rs
Comment (ParthDesai) [inline solo-chains/runtime/dancelight/src/lib.rs:221]: Also, you might not need this as well. Since it looks like the only use for outbound queue pallet for `AggregateMessageOrigin` is to enqueue it for later processing by itself in which case the origin is ignored anyway. 

If this requires widespread changes, ignore this comment!

Comment (tmpolaczyk) [reply solo-chains/runtime/dancelight/src/lib.rs:221]: I believe this is for backwards compatibility, before this PR we used the `AggregateMessageOrigin` from the paras_inclusion pallet, which has Ump as index 0. Also we probably will need to support this at some point in the future (for xcm maybe?). So I'm ignoring this

File: solo-chains/runtime/dancelight/src/lib.rs
Comment (tmpolaczyk) [reply solo-chains/runtime/dancelight/src/lib.rs:218]: Typo, I meant to say outbound queue

Comment (tmpolaczyk) [reply solo-chains/runtime/dancelight/src/lib.rs:218]: *fixed*

File: solo-chains/runtime/dancelight/src/lib.rs
Comment (tmpolaczyk) [reply solo-chains/runtime/dancelight/src/lib.rs:221]: *disregard*

File: solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs
Comment (ParthDesai) [inline solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs:35]: I am curious about this exchange rate. Was this copied from the snowbridge config?

Comment (ParthDesai) [reply solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs:35]: *disregard*

**Review by ParthDesai — APPROVED** [review]
Looks good to me. 

Only remaining part would be to define command and review the overall flow but that can be handled in next PR.

File: solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs
Comment (girazoki) [reply solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs:142]: so can you do at least local benchmarks?

Comment (girazoki) [reply solo-chains/runtime/dancelight/src/bridge_to_ethereum_config.rs:142]: ignore

**Pull request merged**
